pusher/pusher.go

```diff
@@ -0,0 +1,219 @@
+package pusher
+
+import (
+	"bytes"
+	"crypto/sha1"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"mime/multipart"
+	"net/http"
+	"path"
+	"strings"
+	"time"
+
+	"github.com/kochman/hotshots/config"
+	"github.com/kochman/hotshots/log"
+)
+
+const photosEndpoint = "/photos"
+const photoIDEndpoint = photosEndpoint + "/ids"
+
+// this is copied/pasted from the server package... should we just import it?
+type photoIDsResponse struct {
+	Success bool     `json:"success"`
+	IDs     []string `json:"ids"`
+}
+
+// same as above... this is copied/pasted from the server package... should we just import it?
+type photoResponse struct {
+	Success bool `json:"success"`
+	// Status  Status `json:"status"`
+	NewID string `json:"new_id"`
+}
+
+type Pusher struct {
+	cfg               config.Config
+	camera            camera
+	filenameToPhotoID map[string]string
+}
+
+func New(cfg *config.Config) (*Pusher, error) {
+	p := &Pusher{
+		cfg:               *cfg,
+		camera:            newGphotoCamera(),
+		filenameToPhotoID: map[string]string{},
+	}
+
+	return p, nil
+}
+
+func (p *Pusher) Run() {
+
+	ticker := time.NewTicker(time.Second)
+	for range ticker.C {
+		p.generatePhotoIDs()
+
+		photos, err := p.getExistingPhotos()
+		if err != nil {
+			log.WithError(err).Error("unable to get existing photos")
+			continue
+		}
+		log.Infof("%+v", photos)
+
+		toUpload := []string{}
+		for filename, id := range p.filenameToPhotoID {
+			found := false
+			for _, existing := range photos {
+				if id == existing {
+					found = true
+					break
+				}
+			}
+			if !found {
+				toUpload = append(toUpload, filename)
+			}
+		}
+
+		log.Infof("%+v", toUpload)
+		for _, filename := range toUpload {
+			// only transfer JPEGs
+			if strings.ToLower(path.Ext(filename)) != ".jpg" {
+				continue
+			}
+
+			b, err := p.camera.getFile(filename)
+			if err != nil {
+				log.WithError(err).Error("unable to get file")
+				continue
+			}
+
+			err = p.uploadPhoto(b)
+			if err != nil {
+				log.WithError(err).Errorf("unable to upload %s", filename)
+				continue
+			}
+			log.Infof("uploaded photo %s", filename)
+		}
+	}
+}
+
+func (p *Pusher) getExistingPhotos() ([]string, error) {
+	c := &http.Client{
+		Timeout: 5 * time.Second,
+	}
+
+	req, err := http.NewRequest("GET", p.cfg.PusherDestURL+photoIDEndpoint, nil)
+	if err != nil {
+		return []string{}, err
+	}
+
+	resp, err := c.Do(req)
+	if err != nil {
+		return []string{}, err
+	}
+
+	var idsResp photoIDsResponse
+	dec := json.NewDecoder(resp.Body)
+	err = dec.Decode(&idsResp)
+	if err != nil {
+		return []string{}, err
+	}
+
+	return idsResp.IDs, nil
+}
+
+func (p *Pusher) uploadPhoto(photo []byte) error {
+	c := &http.Client{
+		Timeout: 30 * time.Second,
```

Comment:
This is a fairly generous timeout. Are we expecting this to be going over slow internet?